File: js/engine/answerTypes.js
// Canonical answer shape returned by every question handler and by the engine.
// type: 'yes' | 'no' | 'sometimes' | 'unknown'
//
// SECURITY / GAME-INTEGRITY RULE: answers returned from here must never be able to
// identify or narrow down the secret character. That means:
//   - No character name, ever.
//   - No per-character `notes` text (character.notes exists for the POST-GAME
//     reveal screen only — see js/ui/result.js — and must never be threaded
//     through to an in-round answer).
//   - The only optional clarifying text allowed is a generic, question-level
//     hint (e.g. "This depends on how the term is defined.") that is identical
//     for every character who could produce that answer, so it can never single
//     one out. Pass it as `hint` from questionBank.js, never derived per-character.

export const DISPLAY_LABEL = {
  yes: 'Yes',
  no: 'No',
  sometimes: 'Sometimes',
  unknown: 'Not enough information',
};

export function makeAnswer(type, { hint } = {}) {
  const label = DISPLAY_LABEL[type] || DISPLAY_LABEL.unknown;
  return {
    type,
    label,
    hint: hint || null,
    text: hint ? `${label}. ${hint}` : `${label}.`,
  };
}

// Maps a tri-state fact field ('yes' | 'no' | 'debated') to an answer.
// `sometimesHint`, if given, must be a generic, question-level sentence — never
// anything derived from the specific character.
export function answerFromTriState(value, { sometimesHint } = {}) {
  if (value === 'yes') return makeAnswer('yes');
  if (value === 'no') return makeAnswer('no');
  if (value === 'debated') return makeAnswer('sometimes', { hint: sometimesHint });
  return makeAnswer('unknown');
}

// Maps a boolean field to a plain yes/no answer.
export function answerFromBoolean(value) {
  return makeAnswer(value ? 'yes' : 'no');
}

// Maps a 'yes' | 'no' | 'unknown' field (married, had_children, violent_death) to an answer.
export function answerFromYesNoUnknown(value) {
  if (value === 'yes') return makeAnswer('yes');
  if (value === 'no') return makeAnswer('no');
  return makeAnswer('unknown');
}

// Maps a before/during/after/spans landmark field, relative to the asked direction.
export function answerFromTimeline(value, direction, { sometimesHint } = {}) {
  if (value === 'spans' || value === 'during') return makeAnswer('sometimes', { hint: sometimesHint });
  if (value === 'na') return makeAnswer('unknown');
  const isBefore = value === 'before';
  const matches = direction === 'before' ? isBefore : !isBefore;
  return makeAnswer(matches ? 'yes' : 'no');
}

File: server/answer_types.py
"""Canonical answer shape, mirroring js/engine/answerTypes.js exactly.

GAME-INTEGRITY RULE: an answer must never be able to identify or narrow down the
secret character. `hint` may ONLY be a fixed, generic, question-level sentence
(identical for every character who could produce that answer) — never anything
derived from the specific character. See question_bank.py for where hints come from.
"""

DISPLAY_LABEL = {
    "yes": "Yes",
    "no": "No",
    "sometimes": "Sometimes",
    "unknown": "Not enough information",
}


def make_answer(answer_type, hint=None):
    label = DISPLAY_LABEL.get(answer_type, DISPLAY_LABEL["unknown"])
    return {
        "type": answer_type,
        "label": label,
        "hint": hint,
        "text": f"{label}. {hint}" if hint else f"{label}.",
    }


def answer_from_tri_state(value, sometimes_hint=None):
    """value: 'yes' | 'no' | 'debated'"""
    if value == "yes":
        return make_answer("yes")
    if value == "no":
        return make_answer("no")
    if value == "debated":
        return make_answer("sometimes", sometimes_hint)
    return make_answer("unknown")


def answer_from_boolean(value):
    return make_answer("yes" if value else "no")


def answer_from_yes_no_unknown(value):
    """value: 'yes' | 'no' | 'unknown'"""
    if value == "yes":
        return make_answer("yes")
    if value == "no":
        return make_answer("no")
    return make_answer("unknown")


def answer_from_timeline(value, direction, sometimes_hint=None):
    """value: 'before' | 'during' | 'after' | 'spans' | 'na'; direction: 'before' | 'after'"""
    if value in ("spans", "during"):
        return make_answer("sometimes", sometimes_hint)
    if value == "na":
        return make_answer("unknown")
    is_before = value == "before"
    matches = is_before if direction == "before" else not is_before
    return make_answer("yes" if matches else "no")
